packages/block/src/block.ts

```diff
@@ -299,11 +316,76 @@ export class Block {
     }
   }
 
-  private _validateUncleHeader(uncleHeader: BlockHeader, blockchain: Blockchain) {
-    // TODO: Validate that the uncle header hasn't been included in the blockchain yet.
-    // This is not possible in ethereumjs-blockchain since this PR was merged:
-    // https://github.com/ethereumjs/ethereumjs-blockchain/pull/47
-    const height = this.header.number
-    return uncleHeader.validate(blockchain, height)
+  private async _validateUncleHeaders(uncleHeaders: BlockHeader[], blockchain: Blockchain) {
+    if (uncleHeaders.length == 0) {
+      return
+    }
+
+    // validate the header
+    for (let i = 0; i < uncleHeaders.length; i++) {
+      const header = uncleHeaders[i]
+      await header.validate(blockchain, this.header.number)
+    }
+
+    const canonicalBlockMap: Block[] = []
+
+    let lowestUncleNumber = this.header.number.clone()
+
+    uncleHeaders.map((header) => {
+      if (header.number.lt(lowestUncleNumber)) {
+        lowestUncleNumber = header.number.clone()
+      }
+    })
+
+    const getBlocks = this.header.number.clone().sub(lowestUncleNumber).addn(1).toNumber()
```

Comment:
If the `lowestUncleNumber` here is e.g. 1 for whatever reason the following loop would go and re-read the whole blockchain. This can't be right I would assume? 🤔

Reply:
I am not seeing why this would read the whole blockchain. Please note that in validating the headers, it is checked that the uncle is at most 8 blocks older, so this *should* put an upper bound of the amount of blocks being read.

Reply:
Thanks, I wasn't aware of this check in uncle header validation.